In [39]:
import pandas as pd

In [40]:
df = pd.read_csv('upstart-loan-data.csv')

In [41]:
df

Unnamed: 0 Loan ID Loan Grade  Term (months) Origination Date  \
0          15758  L17471          C             36       2015-11-04   
1          46818  L49499          B             60       2016-12-19   
2           7461  L28976          C             60       2016-03-23   
3            191    L286          C             36       2014-06-10   
4          46967  L53569          B             36       2017-02-01   
...          ...     ...        ...            ...              ...   
9995       22698  L27763          A             36       2016-03-07   
9996        6891   L5683          B             36       2015-04-06   
9997       61625  L67159          B             36       2017-05-16   
9998       15024  L12888          A             36       2015-09-10   
9999       29531  L30309          E             36       2016-04-11   

      Interest Rate    APR  Origination Fee  Loan Size  Outstanding Principal  \
0             12.56  16.94           492.00     8200.0                4733.79   
1             15.98  18.64          1400.00    25000.0               22987.76   
2             18.02  20.44           325.00     6500.0                5296.60   
3             14.15  17.80           250.00     5000.0                   0.00   
4             10.02  13.43           239.10     5000.0                4388.82   
...             ...    ...              ...        ...                    ...   
9995           7.92   9.88           436.13    15400.0                8980.34   
9996          10.93  14.51           800.00    16000.0                   0.00   
9997           9.07  12.80          1363.18    26000.0               24733.78   
9998           7.23  10.72           320.00     6400.0                   0.00   
9999          26.01  29.93          1685.00    33700.0               24462.63   

      Days Late Delinquency Status  Charged Off Amount Charged Off Date  
0            89         Delinquent                 NaN              NaN  
1             0            Current                 NaN              NaN  
2             0            Current                 NaN              NaN  
3             0            Matured                 NaN              NaN  
4             0            Current                 NaN              NaN  
...         ...                ...                 ...              ...  
9995          0            Current                 NaN              NaN  
9996          0            Matured                 NaN              NaN  
9997          0            Current                 NaN              NaN  
9998          0            Matured                 NaN              NaN  
9999         51         Delinquent                 NaN              NaN  

[10000 rows x 14 columns]

### Question 1

Drop all columns except for these... <b>['Loan Grade', 'Term (months)', 'Interest Rate', 'APR', 'Origination Fee', 'Loan Size', 'Outstanding Principal', 'Days Late', 'Delinquency Status']</b>

In [42]:
#df.drop(['Loan Grade', 'Term (months)', 'Interest Rate', 'APR', 'Origination Fee', 'Loan Size',
#         'Outstanding Principal', 'Days Late', 'Delinquency Status'], axis = 1, inplace = True)

In [43]:
df.drop(['Unnamed: 0', 'Loan ID', 'Origination Date', 'Charged Off Amount', 'Charged Off Date'],
        axis = 1, inplace = True)

In [44]:
df

Loan Grade  Term (months)  Interest Rate    APR  Origination Fee  \
0             C             36          12.56  16.94           492.00   
1             B             60          15.98  18.64          1400.00   
2             C             60          18.02  20.44           325.00   
3             C             36          14.15  17.80           250.00   
4             B             36          10.02  13.43           239.10   
...         ...            ...            ...    ...              ...   
9995          A             36           7.92   9.88           436.13   
9996          B             36          10.93  14.51           800.00   
9997          B             36           9.07  12.80          1363.18   
9998          A             36           7.23  10.72           320.00   
9999          E             36          26.01  29.93          1685.00   

      Loan Size  Outstanding Principal  Days Late Delinquency Status  
0        8200.0                4733.79         89         Delinquent  
1       25000.0               22987.76          0            Current  
2        6500.0                5296.60          0            Current  
3        5000.0                   0.00          0            Matured  
4        5000.0                4388.82          0            Current  
...         ...                    ...        ...                ...  
9995    15400.0                8980.34          0            Current  
9996    16000.0                   0.00          0            Matured  
9997    26000.0               24733.78          0            Current  
9998     6400.0                   0.00          0            Matured  
9999    33700.0               24462.63         51         Delinquent  

[10000 rows x 9 columns]

### Interest Rates vs. APR

You may have or you will come across terms like interest rates and APR when you go out to get loans for homes, auto and in this instance (personal loans from Upstart, a lending company), as a borrower trying to get unsecured credit. The difference between interest rate and APR in the context of refinancing or taking out a mortgage (for a home) is...

- Interest rate refers to the annual cost of a loan to a borrower and is expressed as a percentage.
- APR is the annual cost of a loan to a borrower — including fees. Like an interest rate, the APR is expressed as a percentage. Unlike an interest rate however, it includes other charges or fees such as mortgage insurance, most closing costs, discount points and loan origination fees.

But what's the difference between unsecured and secured loan? If you borrow money to buy a home, say from a bank and if you were to skip paying on that loan, guess what the bank does? It takes away your home. So the bank is making a loan based on the fact that in worst-case scenarios, they have an asset (home) they can go after. That's secured credit and hence generally carry a lower interest rate.

Unsecured credit would be like a bank lending money to you on your credit card. If you don't pay, they will cancel your credit and destroy your credit score but that is it. They can't technically come after your other assets. That's a form of unsecured credit.

### Question 2

Now that you know the difference, drop the <b>Interest Rate</b> column as well and rename the column <b>Term (months)</b> to <b>Term</b>. Also check if all attributes are in the correct format (integers, float etc). If not, change them into the right format.

In [45]:
df.drop('Interest Rate', axis = 1, inplace = True)
df

Loan Grade  Term (months)    APR  Origination Fee  Loan Size  \
0             C             36  16.94           492.00     8200.0   
1             B             60  18.64          1400.00    25000.0   
2             C             60  20.44           325.00     6500.0   
3             C             36  17.80           250.00     5000.0   
4             B             36  13.43           239.10     5000.0   
...         ...            ...    ...              ...        ...   
9995          A             36   9.88           436.13    15400.0   
9996          B             36  14.51           800.00    16000.0   
9997          B             36  12.80          1363.18    26000.0   
9998          A             36  10.72           320.00     6400.0   
9999          E             36  29.93          1685.00    33700.0   

      Outstanding Principal  Days Late Delinquency Status  
0                   4733.79         89         Delinquent  
1                  22987.76          0            Current  
2                   5296.60          0            Current  
3                      0.00          0            Matured  
4                   4388.82          0            Current  
...                     ...        ...                ...  
9995                8980.34          0            Current  
9996                   0.00          0            Matured  
9997               24733.78          0            Current  
9998                   0.00          0            Matured  
9999               24462.63         51         Delinquent  

[10000 rows x 8 columns]

In [46]:
df.rename(columns={"Term (months)": "Term"}, inplace = True)
df

Loan Grade  Term    APR  Origination Fee  Loan Size  \
0             C    36  16.94           492.00     8200.0   
1             B    60  18.64          1400.00    25000.0   
2             C    60  20.44           325.00     6500.0   
3             C    36  17.80           250.00     5000.0   
4             B    36  13.43           239.10     5000.0   
...         ...   ...    ...              ...        ...   
9995          A    36   9.88           436.13    15400.0   
9996          B    36  14.51           800.00    16000.0   
9997          B    36  12.80          1363.18    26000.0   
9998          A    36  10.72           320.00     6400.0   
9999          E    36  29.93          1685.00    33700.0   

      Outstanding Principal  Days Late Delinquency Status  
0                   4733.79         89         Delinquent  
1                  22987.76          0            Current  
2                   5296.60          0            Current  
3                      0.00          0            Matured  
4                   4388.82          0            Current  
...                     ...        ...                ...  
9995                8980.34          0            Current  
9996                   0.00          0            Matured  
9997               24733.78          0            Current  
9998                   0.00          0            Matured  
9999               24462.63         51         Delinquent  

[10000 rows x 8 columns]

In [50]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 10000 entries, 0 to 9999
Data columns (total 8 columns):
 #   Column                 Non-Null Count  Dtype  
---  ------                 --------------  -----  
 0   Loan Grade             10000 non-null  object 
 1   Term                   10000 non-null  int64  
 2   APR                    10000 non-null  float64
 3   Origination Fee        10000 non-null  float64
 4   Loan Size              10000 non-null  float64
 5   Outstanding Principal  10000 non-null  float64
 6   Days Late              10000 non-null  int64  
 7   Delinquency Status     10000 non-null  object 
dtypes: float64(4), int64(2), object(2)
memory usage: 625.1+ KB


### Question 3

Describe in a few words what type of variables are <b>Loan Grade</b>, <b>Term</b>, <b>Days Late</b> and <b>Delinquency Status</b> and why.

- Loan Grade:        The category of how risky it is that the loan will not be repaid and the outstanding balance can be recovered (Ordinal Letter Grade. From best to worse: AAA, AA, A, B, C, D, E)
- Term:              The duration of the loan in months (Discrete Values. Here: 36, 60. In reality more values are possible)
- Days Late:         The maximum of days, a debtor was late to pay an installment (Discrete Values: integer number)
- Delinqency Status: Has the debtor delayed/stopped paying installments (Nominal Values, no order)

### Question 4

Compute the average by <b>Loan Grade</b> for all columns and give a few remarks on what you observe for the <b>APR</b> and <b>Days Late</b> columns.

In [52]:
df.groupby(by='Loan Grade').mean()

Term        APR  Origination Fee     Loan Size  \
Loan Grade                                                        
A           40.481858  11.226493       608.045951  12282.244799   
AA          45.478060   8.464965       398.056490  13238.337182   
AAA         48.000000   6.518333       123.333333  12333.333333   
B           43.276596  14.380888       634.248265  12351.097074   
C           39.492491  17.312299       637.277255  11287.053340   
D           41.292308  21.407568       713.364571  12425.421245   
E           48.665546  27.348857       589.105403   9962.857143   

            Outstanding Principal  Days Late  
Loan Grade                                    
A                     6428.668974   6.161587  
AA                    6899.800139   2.766744  
AAA                   9117.633333   0.000000  
B                     7739.511007  10.098404  
C                     7159.271932  15.167271  
D                     6545.772066  39.327473  
E                     7470.002361  18.111765

Remarks:
1. The order of 'Loan Grade' is not in order: 'AAA' is higher then 'AA' is higher then 'A' is higher then 'B' .. 
'Loan Grade' E has the highest APR, but not the highest 'Days Late'. I assume the grade E loans are unsecured
2. The better the 'Loan Grade' the lower the APR
3. The higher the average of 'Days Late', the lower 'Loan Grade' and the higher the 'APR'

### Question 5

Verify if there are any missing values in the dataset. Confirm that by applying the <b>size()</b> aggregation method on any of the columns.

In [54]:
df.groupby(by='Loan Grade').size()

Loan Grade
A      2067
AA      433
AAA       6
B      3008
C      1931
D      1365
E      1190
dtype: int64

In [61]:
df.groupby(by='Loan Grade')['Term'].size()

Loan Grade
A      2067
AA      433
AAA       6
B      3008
C      1931
D      1365
E      1190
Name: Term, dtype: int64

In [62]:
df.groupby(by='Loan Grade')['APR'].size()

Loan Grade
A      2067
AA      433
AAA       6
B      3008
C      1931
D      1365
E      1190
Name: APR, dtype: int64

In [63]:
df.groupby(by='Loan Grade')['Origination Fee'].size()

Loan Grade
A      2067
AA      433
AAA       6
B      3008
C      1931
D      1365
E      1190
Name: Origination Fee, dtype: int64

In [64]:
df.groupby(by='Loan Grade')['Loan Size'].size()

Loan Grade
A      2067
AA      433
AAA       6
B      3008
C      1931
D      1365
E      1190
Name: Loan Size, dtype: int64

In [65]:
df.groupby(by='Loan Grade')['Outstanding Principal'].size()

Loan Grade
A      2067
AA      433
AAA       6
B      3008
C      1931
D      1365
E      1190
Name: Outstanding Principal, dtype: int64

In [66]:
df.groupby(by='Loan Grade')['Days Late'].size()

Loan Grade
A      2067
AA      433
AAA       6
B      3008
C      1931
D      1365
E      1190
Name: Days Late, dtype: int64

In [67]:
df.groupby(by='Loan Grade')['Delinquency Status'].size()

Loan Grade
A      2067
AA      433
AAA       6
B      3008
C      1931
D      1365
E      1190
Name: Delinquency Status, dtype: int64

In [69]:
df.groupby(by='Loan Grade')['Term', 'APR', 'Origination Fee', 'Loan Size', 'Outstanding Principal', 
                            'Days Late', 'Delinquency Status'].size()

/Users/marco/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  """Entry point for launching an IPython kernel.


Loan Grade
A      2067
AA      433
AAA       6
B      3008
C      1931
D      1365
E      1190
dtype: int64

In [55]:
df.groupby(by='Loan Grade').count()

Term   APR  Origination Fee  Loan Size  Outstanding Principal  \
Loan Grade                                                                  
A           2067  2067             2067       2067                   2067   
AA           433   433              433        433                    433   
AAA            6     6                6          6                      6   
B           3008  3008             3008       3008                   3008   
C           1931  1931             1931       1931                   1931   
D           1365  1365             1365       1365                   1365   
E           1190  1190             1190       1190                   1190   

            Days Late  Delinquency Status  
Loan Grade                                 
A                2067                2067  
AA                433                 433  
AAA                 6                   6  
B                3008                3008  
C                1931                1931  
D                1365                1365  
E                1190                1190

### Question 6

Display the count by <b>Loan Grade</b> of all the customers in the dataset. Also display the <b>max</b>, <b>min</b> and <b>mean</b> by <b>Loan Grade</b> of <b>Outstanding Principal</b>. Which loan grade has the least number of customers? Which loan grade has the lowest spread? Does the rating of the lowest spread customers conform to what you would expect if you were the lender?

In [73]:
df.groupby(by='Loan Grade')['Outstanding Principal'].agg(['count', 'min', 'max', 'mean'])

count  min       max         mean
Loan Grade                                   
A            2067  0.0  49351.47  6428.668974
AA            433  0.0  49332.70  6899.800139
AAA             6  0.0  21720.21  9117.633333
B            3008  0.0  49380.34  7739.511007
C            1931  0.0  50000.00  7159.271932
D            1365  0.0  49436.30  6545.772066
E            1190  0.0  50000.00  7470.002361

Remarks:
1. The 'Loan Grade' AAA has the lowest number of customers
2. The 'Loan Grade' AAA has the lowest spread of 21720.21
3. If the 'Loan Grade' is AAA, you would expect that the lendors is willing to lend more money. Maybe, because the spread is so low, the grading of the loan is higher (easy to be paid back). 

### Question 7

Does <b>Loan Size</b> provide any information about <b>Loan Grade</b>? Is a high risk borrower more likely to take bigger loans than a low risk borrower based on the data you have? Define in a few words what risk is to a lender? And to a borrower?

In [77]:
df.groupby(by='Loan Grade')['Loan Size'].agg(['count', 'min', 'max', 'mean'])

count     min      max          mean
Loan Grade                                      
A            2067  1000.0  50000.0  12282.244799
AA            433  1000.0  50000.0  13238.337182
AAA             6  5000.0  25000.0  12333.333333
B            3008  1000.0  50000.0  12351.097074
C            1931  1000.0  50000.0  11287.053340
D            1365  1000.0  50000.0  12425.421245
E            1190  1000.0  50000.0   9962.857143

Remarks:
1. Loan Grade 'AAA' is limited to 25,000, while all other grades can be obtained up to 50,000
2. Loans of the grade 'E' are more likely to be used for smaller amounts compaired with other loan grades. 

Lenders Risk:
    - If a lender lends money to a borrower, the ideal case is a constant pay from the borrower to the lender.
    - If the borrower delays paying, the lender can not use the money to lend it to somebody else (opportunity loss). 
    - If the borrower stops paying/becomes delinquent, the lender needs to take action to recover the remainder of the loan: sell the loan to a debt collector, sell/auction the asset/security/collateral with administrative costs. 
    
Borrowers Risk:
    - If a borrower borrows money from a lender, the borrower can 'invest' the loan and use it for an asset/security or cover a temporary financial set back. 
    - If the borrower does not pay the installments on time, the lender has to opportunity to cancel the loan and take possession of the asset/security/collateral and influence the borrowers future borrowing rating (credit score). 
    - If the loan was used to cover for a temporary financial set back of a business, this could force the business into bankruptcy (unemployment for employees) as well as credit score of the borrower. 

### Question 8

Display the count by <b>Loan Grade</b> of all the customers in the dataset and call that column <b>Count</b>. Also display the <b>max</b>, <b>min</b> and <b>mean</b> by <b>Loan Grade</b> of <b>Outstanding Principal</b> and call them <b>Max Balance</b>, <b>Min Balance</b> and <b>Average Balance</b>. 

In [89]:
df.groupby(by='Loan Grade')['Outstanding Principal'].agg([('Count', 'count'), ('Max Balance','max'), 
                                                          ('Min Balance', 'min'), ('Average Balance,','mean')
                                                         ])

Count  Max Balance  Min Balance  Average Balance,
Loan Grade                                                   
A            2067     49351.47          0.0       6428.668974
AA            433     49332.70          0.0       6899.800139
AAA             6     21720.21          0.0       9117.633333
B            3008     49380.34          0.0       7739.511007
C            1931     50000.00          0.0       7159.271932
D            1365     49436.30          0.0       6545.772066
E            1190     50000.00          0.0       7470.002361

### Question 9

Write a code that takes the <b>Term</b> column and replace the values within that to <b>0</b> for <b>36</b> months and <b>1</b> for <b>60</b> months using a loop. Do the same using a lambda function. And then, using the <b>get_dummies</b> method (you might have to write additional code to drop redundant columns).

In [171]:
df.Term.unique()

array([36, 60])

In [172]:
df.Term.value_counts()

36    7318
60    2682
Name: Term, dtype: int64

#### functional approach

In [173]:
def dummy(column):
    if column == 36:
        return 0
    elif column == 60:
        return 1
    else:
        print("ERROR: An unexpected value: ", column)
        exit(1)

In [174]:
temp = df.copy()

In [175]:
temp.Term = temp.Term.apply(dummy)

In [176]:
temp.Term.values

array([0, 1, 1, ..., 0, 0, 0])

In [177]:
temp.Term.unique()

array([0, 1])

In [178]:
temp.Term.value_counts()

0    7318
1    2682
Name: Term, dtype: int64

#### lambda approach

In [195]:
temp = df.copy()

In [196]:
temp.Term=temp.Term.apply(lambda x: 0 if x == 36 else 1)
temp

Loan Grade  Term    APR  Origination Fee  Loan Size  \
0             C     0  16.94           492.00     8200.0   
1             B     1  18.64          1400.00    25000.0   
2             C     1  20.44           325.00     6500.0   
3             C     0  17.80           250.00     5000.0   
4             B     0  13.43           239.10     5000.0   
...         ...   ...    ...              ...        ...   
9995          A     0   9.88           436.13    15400.0   
9996          B     0  14.51           800.00    16000.0   
9997          B     0  12.80          1363.18    26000.0   
9998          A     0  10.72           320.00     6400.0   
9999          E     0  29.93          1685.00    33700.0   

      Outstanding Principal  Days Late Delinquency Status  
0                   4733.79         89         Delinquent  
1                  22987.76          0            Current  
2                   5296.60          0            Current  
3                      0.00          0            Matured  
4                   4388.82          0            Current  
...                     ...        ...                ...  
9995                8980.34          0            Current  
9996                   0.00          0            Matured  
9997               24733.78          0            Current  
9998                   0.00          0            Matured  
9999               24462.63         51         Delinquent  

[10000 rows x 8 columns]

In [197]:
temp.Term.unique()

array([0, 1])

In [198]:
temp.Term.value_counts()

0    7318
1    2682
Name: Term, dtype: int64

#### dummy approach

In [199]:
temp = df.copy()

In [200]:
temp=pd.get_dummies(temp, columns=['Term'])
temp

Loan Grade    APR  Origination Fee  Loan Size  Outstanding Principal  \
0             C  16.94           492.00     8200.0                4733.79   
1             B  18.64          1400.00    25000.0               22987.76   
2             C  20.44           325.00     6500.0                5296.60   
3             C  17.80           250.00     5000.0                   0.00   
4             B  13.43           239.10     5000.0                4388.82   
...         ...    ...              ...        ...                    ...   
9995          A   9.88           436.13    15400.0                8980.34   
9996          B  14.51           800.00    16000.0                   0.00   
9997          B  12.80          1363.18    26000.0               24733.78   
9998          A  10.72           320.00     6400.0                   0.00   
9999          E  29.93          1685.00    33700.0               24462.63   

      Days Late Delinquency Status  Term_36  Term_60  
0            89         Delinquent        1        0  
1             0            Current        0        1  
2             0            Current        0        1  
3             0            Matured        1        0  
4             0            Current        1        0  
...         ...                ...      ...      ...  
9995          0            Current        1        0  
9996          0            Matured        1        0  
9997          0            Current        1        0  
9998          0            Matured        1        0  
9999         51         Delinquent        1        0  

[10000 rows x 9 columns]

In [201]:
temp.drop('Term_36', axis = 1, inplace = True)
temp.rename(columns={"Term_60": "Term"}, inplace = True)
temp

Loan Grade    APR  Origination Fee  Loan Size  Outstanding Principal  \
0             C  16.94           492.00     8200.0                4733.79   
1             B  18.64          1400.00    25000.0               22987.76   
2             C  20.44           325.00     6500.0                5296.60   
3             C  17.80           250.00     5000.0                   0.00   
4             B  13.43           239.10     5000.0                4388.82   
...         ...    ...              ...        ...                    ...   
9995          A   9.88           436.13    15400.0                8980.34   
9996          B  14.51           800.00    16000.0                   0.00   
9997          B  12.80          1363.18    26000.0               24733.78   
9998          A  10.72           320.00     6400.0                   0.00   
9999          E  29.93          1685.00    33700.0               24462.63   

      Days Late Delinquency Status  Term  
0            89         Delinquent     0  
1             0            Current     1  
2             0            Current     1  
3             0            Matured     0  
4             0            Current     0  
...         ...                ...   ...  
9995          0            Current     0  
9996          0            Matured     0  
9997          0            Current     0  
9998          0            Matured     0  
9999         51         Delinquent     0  

[10000 rows x 8 columns]

In [202]:
temp.Term.unique()

array([0, 1], dtype=uint8)

In [203]:
temp.Term.value_counts()

0    7318
1    2682
Name: Term, dtype: int64

### Question 10

<p>Ask an interesting question of this data set and provide a solution to answer that.</p>
<p>The goal is to help teach your fellow students all possible questions we can collectively ask of this dataset. Your question should be as clear as possible and as short as possible. Try to avoid asking questions that are too trivial or obvious.</p>

<p>This is a bonus point question. The only way not to get full points are if you do the following:
<ul>
<li>You do not perform this task.
<li>You do not include a solution. 
</ul>
</p>
<p>If the solution you provide is incorrect, you will still receive full points. But you must make an honest effort to get it right.
</p>

#### The Question

Find the mean for all columns grouped by Loan Grade and Term (Hierarchical Indexing). How does the Term influence the APR and Loan Size? What is the difference in APR and Loan Size? 

In [208]:
df.groupby(by=['Loan Grade', 'Term']).mean()

APR  Origination Fee     Loan Size  \
Loan Grade Term                                             
A          36    10.859560       569.408745  11860.916121   
           60    12.824456       776.307979  14117.098446   
AA         36     7.473130       324.119122  12374.427481   
           60     9.984620       511.340643  14561.988304   
AAA        36     6.166667        56.666667   5666.666667   
           60     6.870000       190.000000  19000.000000   
B          36    13.776713       571.419165  11496.517176   
           60    15.769430       778.644967  14315.131579   
C          36    17.197206       609.172861  10906.666667   
           60    17.988114       802.303060  13520.640569   
D          36    21.914727       681.345470  12115.225564   
           60    19.614817       826.548372  13521.926910   
E          36    29.042260       508.140694   9020.284698   
           60    25.833424       661.561083  10806.369427   

                 Outstanding Principal  Days Late  
Loan Grade Term                                    
A          36              5049.689709   7.560381  
           60             12434.016503   0.069948  
AA         36              2819.678206   4.480916  
           60             13151.215029   0.140351  
AAA        36              1023.250000   0.000000  
           60             17212.016667   0.000000  
B          36              5741.468597  13.114504  
           60             12331.503213   3.166667  
C          36              6232.457988  17.225455  
           60             12601.417865   3.081851  
D          36              4781.837660  50.125000  
           60             12781.075083   1.159468  
E          36              4157.603363  37.784698  
           60             10434.282994   0.506369

In [318]:
import numpy as np

In [228]:
df['Loan Grade'].unique()

array(['C', 'B', 'D', 'A', 'E', 'AA', 'AAA'], dtype=object)

In [241]:
term_df=df.groupby(by=['Loan Grade','Term'], as_index = False)['APR', 'Loan Size'].mean()
term_df

/Users/marco/opt/anaconda3/lib/python3.7/site-packages/ipykernel_launcher.py:1: FutureWarning: Indexing with multiple keys (implicitly converted to a tuple of keys) will be deprecated, use a list instead.
  """Entry point for launching an IPython kernel.


Loan Grade  Term        APR     Loan Size
0           A    36  10.859560  11860.916121
1           A    60  12.824456  14117.098446
2          AA    36   7.473130  12374.427481
3          AA    60   9.984620  14561.988304
4         AAA    36   6.166667   5666.666667
5         AAA    60   6.870000  19000.000000
6           B    36  13.776713  11496.517176
7           B    60  15.769430  14315.131579
8           C    36  17.197206  10906.666667
9           C    60  17.988114  13520.640569
10          D    36  21.914727  12115.225564
11          D    60  19.614817  13521.926910
12          E    36  29.042260   9020.284698
13          E    60  25.833424  10806.369427

In [243]:
# loan grades
term_df['Loan Grade'].unique()

array(['A', 'AA', 'AAA', 'B', 'C', 'D', 'E'], dtype=object)

In [245]:
# terms
term_df['Term'].unique()

array([36, 60])

In [258]:
term_df.loc[:, 'Loan Grade'] == 'AA'

0     False
1     False
2      True
3      True
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
Name: Loan Grade, dtype: bool

In [259]:
term_df.loc[:, 'Term'] == 36 

0      True
1     False
2      True
3     False
4      True
5     False
6      True
7     False
8      True
9     False
10     True
11    False
12     True
13    False
Name: Term, dtype: bool

In [282]:
(term_df.loc[:, 'Loan Grade'] == 'AA') & (term_df.loc[:, 'Term'] == 36)

0     False
1     False
2      True
3     False
4     False
5     False
6     False
7     False
8     False
9     False
10    False
11    False
12    False
13    False
dtype: bool

In [285]:
term_df.loc[
    (term_df.loc[:, 'Loan Grade'] == 'AA') & (term_df.loc[:, 'Term'] == 36),
    ('APR')
]

2    7.47313
Name: APR, dtype: float64

In [303]:
for grade in term_df['Loan Grade'].unique():
    for term in term_df['Term'].unique():
        print(grade, term)
        print( term_df.loc[
            (term_df.loc[:,'Loan Grade'] == grade) & (term_df.loc[:, 'Term'] == term),
            'APR'
        ])
        print( term_df.loc[
            (term_df.loc[:,'Loan Grade'] == grade) & (term_df.loc[:, 'Term'] == term),
            'Loan Size'
        ])
    print()
        

A 36
0    10.85956
Name: APR, dtype: float64
0    11860.916121
Name: Loan Size, dtype: float64
A 60
1    12.824456
Name: APR, dtype: float64
1    14117.098446
Name: Loan Size, dtype: float64

AA 36
2    7.47313
Name: APR, dtype: float64
2    12374.427481
Name: Loan Size, dtype: float64
AA 60
3    9.98462
Name: APR, dtype: float64
3    14561.988304
Name: Loan Size, dtype: float64

AAA 36
4    6.166667
Name: APR, dtype: float64
4    5666.666667
Name: Loan Size, dtype: float64
AAA 60
5    6.87
Name: APR, dtype: float64
5    19000.0
Name: Loan Size, dtype: float64

B 36
6    13.776713
Name: APR, dtype: float64
6    11496.517176
Name: Loan Size, dtype: float64
B 60
7    15.76943
Name: APR, dtype: float64
7    14315.131579
Name: Loan Size, dtype: float64

C 36
8    17.197206
Name: APR, dtype: float64
8    10906.666667
Name: Loan Size, dtype: float64
C 60
9    17.988114
Name: APR, dtype: float64
9    13520.640569
Name: Loan Size, dtype: float64

D 36
10    21.914727
Name: APR, dtype: float64


Add 2 columns to term_df called APR_diff and Loan_Size_Diff with the rolling difference of 2 succesive 'APR' values  and the roling difference of 2 succesive 'Loan Size' successive values.

In [312]:
term_df['APR_diff']=term_df.APR.diff()

In [313]:
term_df['Loan_Size_Diff']=term_df['Loan Size'].diff()

The difference between 2 different, succeeding loan grades (e.g. APR from A to AA) have no meaning. Set them to NaN

In [321]:
term_df.loc[
    (term_df.loc[:, 'Term'] == 36),
    'APR_diff'
]=np.NaN

In [323]:
term_df.loc[
    (term_df.loc[:, 'Term'] == 36),
    'Loan_Size_Diff'
]=np.NaN

In [324]:
term_df

Loan Grade  Term        APR     Loan Size  APR_diff  Loan_Size_Diff
0           A    36  10.859560  11860.916121       NaN             NaN
1           A    60  12.824456  14117.098446  1.964896     2256.182324
2          AA    36   7.473130  12374.427481       NaN             NaN
3          AA    60   9.984620  14561.988304  2.511490     2187.560823
4         AAA    36   6.166667   5666.666667       NaN             NaN
5         AAA    60   6.870000  19000.000000  0.703333    13333.333333
6           B    36  13.776713  11496.517176       NaN             NaN
7           B    60  15.769430  14315.131579  1.992717     2818.614403
8           C    36  17.197206  10906.666667       NaN             NaN
9           C    60  17.988114  13520.640569  0.790908     2613.973903
10          D    36  21.914727  12115.225564       NaN             NaN
11          D    60  19.614817  13521.926910 -2.299910     1406.701346
12          E    36  29.042260   9020.284698       NaN             NaN
13          E    60  25.833424  10806.369427 -3.208836     1786.084729

#### Question 10 Conclusion

If the term is longer, the APR for high graded loans increases. 
For the Loan Grades D and E, the APR is better for the longer term loans

It the term is longer, the Loan Size is also bigger. 
For AAA loans the Loan Size is significant higher (200%) for longer terms 
(maybe this is due to only 6 out of 10,000 loans fall into this category). 
For D and E loans, the Loan Size is still bigger for longer terms, but is 15% or less. 